sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/LogsIngestionAsyncClient.java

```diff
@@ -0,0 +1,268 @@
+        }
+        return new UploadLogsResult(UploadLogsStatus.SUCCESS, null);
+    }
+
+    private Mono<UploadLogsResponseHolder> uploadToService(String dataCollectionRuleId, String streamName, RequestOptions requestOptions, byte[] bytes) {
+        return service.uploadWithResponse(dataCollectionRuleId, streamName,
+                        BinaryData.fromBytes(bytes), requestOptions)
+                .map(response -> new UploadLogsResponseHolder(UploadLogsStatus.SUCCESS, null))
+                .onErrorResume(HttpResponseException.class,
+                        ex -> Mono.just(new UploadLogsResponseHolder(UploadLogsStatus.FAILURE,
+                                mapToResponseError(ex))));
+    }
+
+    /**
+     * Method to map the exception to {@link ResponseError}.
+     * @param ex the {@link HttpResponseException}.
+     * @return the mapped {@link ResponseError}.
+     */
+    private ResponseError mapToResponseError(HttpResponseException ex) {
+        ResponseError responseError = null;
+        // with DPG clients, the error object is just a LinkedHashMap and should map to the standard error
+        // response structure
+        // https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/data-plane/v1/types.json#L46
+        if (ex.getValue() instanceof LinkedHashMap<?, ?>) {
+            @SuppressWarnings("unchecked")
+            LinkedHashMap<String, Object> errorMap = (LinkedHashMap<String, Object>) ex.getValue();
+            if (errorMap.containsKey("error")) {
+                Object error = errorMap.get("error");
+                if (error instanceof LinkedHashMap<?, ?>) {
+                    @SuppressWarnings("unchecked")
+                    LinkedHashMap<String, String> errorDetails = (LinkedHashMap<String, String>) error;
+                    if (errorDetails.containsKey("code") && errorDetails.containsKey("message")) {
+                        responseError = new ResponseError(errorDetails.get("code"), errorDetails.get("message"));
+                    }
+                }
+            }
+        }
+        return responseError;
+    }
+
+    private UploadLogsResult createResponse(List<UploadLogsResult> results) {
+        boolean allErrors = results.stream().allMatch(result -> result.getStatus() == UploadLogsStatus.FAILURE);
+        if (allErrors) {
+            return new UploadLogsResult(UploadLogsStatus.FAILURE,
+                    results.stream().flatMap(result -> result.getErrors().stream()).collect(Collectors.toList()));
+        }
+
+        boolean anyErrors = results.stream().anyMatch(result -> result.getStatus() == UploadLogsStatus.FAILURE);
+        if (anyErrors) {
+            return new UploadLogsResult(UploadLogsStatus.PARTIAL_FAILURE,
+                    results.stream().filter(result -> result.getStatus() == UploadLogsStatus.FAILURE)
+                            .flatMap(result -> result.getErrors().stream()).collect(Collectors.toList()));
+        }
```

Comment:
Could this all be simplified into a single `for` loop? It looks like FAILURE vs PARTIAL_FAILURE is based on whether all requests were failures or only some, this could be made into a simple if

```java
int failureCount = 0;
List<ResponseError> errors = new ArrayList(); // This could be instantiated when needed but this is just a GitHub comment

for (...) {
  if (result.getStatus() == UploadLogsResult.SUCCESS) {
    continue; // Was successful, nothing to inspect.
  }

  failureCount++; // Increase the failure count.
  errors.add(result.getErrors());
}

if (failureCount == 0) {
  return new UploadLogsResult(UploadLogsStatus.SUCCESS, null);
} else {
  return new UploadLogsResult(failureCount == results.size() ? UploadLogsStatus.FAILURE : UploadLogsStatus.PARTIAL_FAILURE, errors);
}
```